swaybar/tray/item.c

```diff
@@ -493,24 +493,36 @@ uint32_t render_sni(cairo_t *cairo, struct swaybar_output *output, double *x,
 		cairo_destroy(cairo_icon);
 	}
 
-	int padded_size = icon_size + 2*padding;
-	*x -= padded_size;
-	int y = floor((height - padded_size) / 2.0);
+	double descaled_padding = (double)padding / output->scale;
+	double descaled_icon_size = (double)icon_size / output->scale;
+
+	int size = descaled_icon_size + 2 * descaled_padding;
+	*x -= size;
+	int icon_y = floor((output->height - size) / 2.0);
 
 	cairo_operator_t op = cairo_get_operator(cairo);
 	cairo_set_operator(cairo, CAIRO_OPERATOR_OVER);
-	cairo_set_source_surface(cairo, icon, *x + padding, y + padding);
-	cairo_rectangle(cairo, *x, y, padded_size, padded_size);
+
+	cairo_matrix_t scale_matrix;
+	cairo_pattern_t *icon_pattern = cairo_pattern_create_for_surface(icon);
+	// TODO: check cairo_pattern_status for "ENOMEM"
+	cairo_matrix_init_scale(&scale_matrix, output->scale, output->scale);
+	cairo_matrix_translate(&scale_matrix, -(*x + descaled_padding), -(icon_y + descaled_padding));
+	cairo_pattern_set_matrix(icon_pattern, &scale_matrix);
+	cairo_set_source(cairo, icon_pattern);
+	cairo_rectangle(cairo, *x, icon_y, size, size);
```

Comment:
Hmm, I'm not quite sure I understand this call.

The coordinates used here seem to be negated by `cairo_matrix_translate` above? Shouldn't `size` not include the padding?

Reply:
Just as before, the pattern that is setup is painted into the padded region.  Now though, the padding is applied within the pattern to be painted, allowing for pixel-spec.  We want to remove allowing for pixel-spec though, but that was requested as a separate (post-this-fix) change.

I didn't and am not sure I do still fully understand the transform / translation setup, but it _is_ the same technique that `cairo_set_source_surface` does.  I was confused that the source surface needs the full x/y translation (origin setup) in addition to being passed to `cairo_rectangle`, but that's how it works I guess.

Reply:
Hm, reading `_cairo_default_context_set_source_surface`, indeed you're right. Well, let's just merge this for now and see how we can simplify it in follow-up PRs.